src/utils/havoc.js

```diff
@@ -157,12 +168,75 @@ class ObjectValueHavocingVisitor {
     // so that any mutation and property access get tracked after this.
     if (!obj.isHavocedObject()) {
       obj.havoc();
+      if (obj.symbols.size > 0) {
+        throw new FatalError("TODO: Support havocing objects with symbols");
+      }
+      if (obj.unknownProperty !== undefined) {
+        // TODO: Support unknown properties, or throw FatalError.
+        // We have repros, e.g. test/serializer/additional-functions/ArrayConcat.js.
+      }
+      // TODO: We should emit current value and then reset value for all *internal slots*; this will require deep serializer support; or throw FatalError when we detect any non-initial values in internal slots.
+      let realmGenerator = this.realm.generator;
+      for (let [name, propertyBinding] of obj.properties) {
+        // ignore properties with their correct default values
+        if (this.getHeapInspector().canIgnoreProperty(obj, name)) continue;
+
+        let descriptor = propertyBinding.descriptor;
+        if (descriptor === undefined) {
+          // TODO: This happens, e.g. test/serializer/pure-functions/ObjectAssign2.js
+          // If it indeed means deleted binding, should we initialize descriptor with a deleted value?
+          if (realmGenerator !== undefined) realmGenerator.emitPropertyDelete(obj, name);
+        } else {
+          let value = descriptor.value;
+          invariant(
+            value === undefined || value instanceof Value,
+            "cannot be an array because we are not dealing with intrinsics here"
+          );
+          if (value === undefined) {
+            // TODO: Deal with accessor properties
+            // We have repros, e.g. test/serializer/pure-functions/AbstractPropertyObjectKeyAssignment.js
+          } else {
+            invariant(value instanceof Value);
+            if (value instanceof EmptyValue) {
+              if (realmGenerator !== undefined) realmGenerator.emitPropertyDelete(obj, name);
+            } else if (value.mightHaveBeenDeleted()) {
+              throw new FatalError("TODO: Support havocing objects with properties that might have been deleted");
+            } else {
+              if (realmGenerator !== undefined) {
+                let targetDescriptor = this.getHeapInspector().getTargetIntegrityDescriptor(obj);
+                if (
+                  descriptor.writable !== targetDescriptor.writable ||
+                  descriptor.configurable !== targetDescriptor.configurable
+                ) {
+                  realmGenerator.emitDefineProperty(obj, name, descriptor);
+                } else {
+                  realmGenerator.emitPropertyAssignment(obj, name, value);
+                }
+              }
+            }
+            this.realm.recordModifiedProperty(propertyBinding);
+
+            if (isFinalObject) {
+              let clonedDescriptor = Object.assign({}, descriptor);
+              clonedDescriptor.leakedFinalDescriptor = descriptor;
```

Comment:
When we join effects, shouldn't we join this new property on the descriptors too? Otherwise we'll lose this property and have no idea what the original value was on the joined effects?